In [49]:
import os
import numpy as np 
import re
import io
from utilities.features_csv import to_csv
import pandas as pd
from IPython.display import display
from nltk.corpus import stopwords
from nltk import word_tokenize

# Preprocessing

In [50]:
import re 
import csv 
docs = {}
labels = {}
#I only work if you run me in the same folder as the actual text files 

with open('training_labels.csv') as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    line_count = 0
    for row in csv_reader:
        if line_count == 0:
            line_count += 1
        else:
            id = int(row[1])
            score= float(row[2])
            labels[id]=score
            line_count += 1
i = 0
for root, dirs, files in os.walk("corpus/"):
    for file in files:
        if file.endswith(".txt"):
            id = re.sub("[^0-9]", "", file)
            id = int(id) 
            path_file = os.path.join(root,file)
            curdir = path_file
            file = open(curdir, 'r', errors='ignore')
            i+=1
            text = file.read()
            docs[id] = text

        

## Below we create the CSV file with the features
We use the following libraries and tools to develop our features. The code for this can be found under utilities.
* Spacy
* NLTK
* Reg
* Textstat

In [3]:
#Create the csv file --> takes a while
to_csv(docs, labels)

In [51]:
data = pd.read_csv("testing.csv")
display(data.head(n=10))

iD  minor  geo-location  contact_email  vendors  sell_personal  \
0  20481      1             0              0        1              0   
1  20484      0             0              1        0              0   
2    854      0             0              0        0              0   
3  20490      1             0              1        0              0   
4  20493      0             0              0        0              0   
5    173      0             0              1        0              0   
6  11096      0             0              1        0              0   
7  20500      0             0              0        0              0   
8    516      0             1              0        0              0   
9  20506      0             0              0        0              0   

   share_personal  Cookies  fog_index  avg_sentence_length  \
0               0        0  75.245659           183.114146   
1               1        1  58.618557           141.546392   
2               0        0  55.264865           133.162162   
3               0        1  70.405195           171.012987   
4               0        0  81.172242           197.930605   
5               0        1  47.731507           114.328767   
6               0        0  49.727168           119.317919   
7               0        1  92.804233           227.010582   
8               0        1  66.037158           160.092896   
9               0        0  42.903529           102.258824   

   flesch_reading_ease  dale_chall_readability_score  Score  
0                -4.42                          9.08    3.0  
1                37.79                          7.02    3.0  
2                46.30                          6.60    2.0  
3                 7.88                          8.48    3.0  
4               -10.99                          9.82    2.0  
5                65.41                          5.67    3.0  
6                60.35                          5.92    2.0  
7               -48.97                         11.26    2.0  
8                18.96                          7.94    2.0  
9                77.66                          5.07    2.0

# Exploration
For our first exploration we will begin by using all of our features

In [59]:
#Split the data into labels and features
scores = data['Score']

features_raw = data.drop(['Score', 'iD'] , axis = 1)
# features_raw = data.drop("vendors", axis =1)

In [60]:
from sklearn.preprocessing import MinMaxScaler
# Initialize a scaler, then apply it to the features
scaler = MinMaxScaler()
numerical = ['fog_index', 'avg_sentence_length', 'flesch_reading_ease', 'dale_chall_readability_score']
features_log_minmax_transform = pd.DataFrame(data = features_raw)
features_log_minmax_transform[numerical] = scaler.fit_transform(features_raw[numerical])


# Show an example of a record with scaling applied
display(features_log_minmax_transform.head(n = 5))

minor  geo-location  contact_email  vendors  sell_personal  share_personal  \
0      1             0              0        1              0               0   
1      0             0              1        0              0               1   
2      0             0              0        0              0               0   
3      1             0              1        0              0               0   
4      0             0              0        0              0               0   

   Cookies  fog_index  avg_sentence_length  flesch_reading_ease  \
0        0   0.660803             0.660803             0.332320   
1        1   0.418325             0.418325             0.569989   
2        0   0.369417             0.369417             0.617905   
3        1   0.590213             0.590213             0.401577   
4        0   0.747232             0.747232             0.295327   

   dale_chall_readability_score  
0                      0.660400  
1                      0.418331  
2                      0.368978  
3                      0.589894  
4                      0.747356

In [61]:
# Import train_test_split
from sklearn.model_selection import train_test_split
# Split the 'features' and 'income' data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features_log_minmax_transform, 
                                                    scores, 
                                                    test_size = 0.2, 
                                                    random_state = 0)
# Show the results of the split
print("Training set has {} samples.".format(X_train.shape[0]))
print("Testing set has {} samples.".format(X_test.shape[0]))

Training set has 347 samples.
Testing set has 87 samples.


# Random Forrest

In [62]:
from sklearn.ensemble import RandomForestClassifier

forest = RandomForestClassifier(n_estimators=15, max_depth=None,random_state=1, criterion="entropy")
forest.fit(X_train, y_train)

predictions = forest.predict(X_test)
from sklearn.metrics import f1_score
f1_score(y_test, predictions, average='macro')

0.8813727603201287

In [63]:
from sklearn.metrics import classification_report
target_names = ['1', '2', '3', '4', '5']
print(classification_report(y_test, predictions, target_names=target_names))

              precision    recall  f1-score   support

           1       0.76      1.00      0.86        16
           2       0.86      0.86      0.86        44
           3       0.88      0.71      0.79        21
           4       1.00      0.80      0.89         5
           5       1.00      1.00      1.00         1

   micro avg       0.85      0.85      0.85        87
   macro avg       0.90      0.88      0.88        87
weighted avg       0.86      0.85      0.85        87



# Extra Decision Trees

In [70]:
from sklearn.ensemble import ExtraTreesClassifier


ET = ExtraTreesClassifier(n_estimators = 1, random_state = 14, criterion='entropy')
ET.fit(X_train, y_train)
predi = ET.predict(X_test)
f1_score(y_test, predi, average = 'macro')

0.9368496841322929

In [65]:
print(classification_report(y_test, predictions, target_names=target_names))

              precision    recall  f1-score   support

           1       0.76      1.00      0.86        16
           2       0.86      0.86      0.86        44
           3       0.88      0.71      0.79        21
           4       1.00      0.80      0.89         5
           5       1.00      1.00      1.00         1

   micro avg       0.85      0.85      0.85        87
   macro avg       0.90      0.88      0.88        87
weighted avg       0.86      0.85      0.85        87



In [71]:
ET.feature_importances_

array([0.03468525, 0.01681815, 0.16091817, 0.09633278, 0.0040461 ,
       0.05148408, 0.08528065, 0.154515  , 0.12417392, 0.12225961,
       0.1494863 ])

# AdaBoost

In [66]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
tree = DecisionTreeClassifier(max_depth=10, criterion='gini', min_samples_split=5)
ada = AdaBoostClassifier(base_estimator=tree, 
                         n_estimators=50, learning_rate=.2,
                         algorithm='SAMME.R', random_state=4)
ada.fit(X_train, y_train)
predictions_ada = ada.predict(X_test)
f1_score(y_test, predictions_ada, average='micro')

0.8505747126436781

In [67]:
print(classification_report(y_test, predictions_ada, target_names=target_names))

              precision    recall  f1-score   support

           1       0.84      1.00      0.91        16
           2       0.86      0.86      0.86        44
           3       0.79      0.71      0.75        21
           4       1.00      0.80      0.89         5
           5       1.00      1.00      1.00         1

   micro avg       0.85      0.85      0.85        87
   macro avg       0.90      0.88      0.88        87
weighted avg       0.85      0.85      0.85        87



# Decision Tree Classifier

In [19]:
from sklearn.tree import DecisionTreeClassifier
dt  = DecisionTreeClassifier(random_state = 47, criterion='gini', splitter = 'random', max_depth=14)
dt.fit(X_train, y_trbain)
f1_score(y_test, predi, average = 'macro')

0.8842324342324343

In [69]:
#Best performing classifier test on cross validation
from sklearn import model_selection
X = features_log_minmax_transform
y = scores
num_folds = 5
num_instances = len(X)
loocv = model_selection.LeaveOneOut()
model = ExtraTreesClassifier(n_estimators = 1, random_state = 14, criterion='entropy')
results = model_selection.cross_val_score(model, X, y, cv=loocv)
print("Accuracy: %.3f%% (%.3f%%)" % (results.mean()*100.0, results.std()*100.0))

Accuracy: 82.488% (38.007%)


# Results
Our best performing classifier is Extra Decision Trees with an mean f1 score of **93%**.